include/pybind11/cast.h

```diff
@@ -686,13 +694,32 @@ class type_caster_generic {
     }
     void check_holder_compat() {}
 
+    /// Try to load with foreign typeinfo, if available. Used when there is no
+    /// native typeinfo, or when the native one wasn't able to produce a value.
+    PYBIND11_NOINLINE bool try_load_foreign_module_local(handle src) {
+        constexpr auto *local_key = "_pybind11_module_local_typeinfo";
+        const auto pytype = src.get_type();
+        if (!hasattr(pytype, local_key))
+            return false;
+
+        type_info *foreign_typeinfo = reinterpret_borrow<capsule>(getattr(pytype, local_key));
+        if (foreign_typeinfo == typeinfo)
+            return false; // Only consider this foreign loader if actually foreign
```

Comment:
I'm thinking:
- the local type masks the global type, is the correct type_info for the given argument, but the load fails anyway.
- the global type gets checked, which also fails (the global type is not right).
- foreign loaders get checked, but because `typeinfo` has been changed to the global one, we don't short-circuit in `try_load_foreign` and so end up recursing.

It's not immediately obvious to me that the first step is impossible; though it's not immediately obvious to me how I can trigger it in a test, either.